### PR TITLE
Feature/four 10253: The edit password needs to enable according the permission

### PR DESCRIPTION
**Pull request by luNunezProcessmaker**
## Issue & Reproduction Steps
Fall-23-–-Figma (6)

## Related Tickets & Packages
- https://processmaker.atlassian.net/browse/FOUR-10253

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

ci:next

### CODE REVIEW

File: ProcessMaker/Http/Controllers/Api/UserController.php [hunk 1]
```diff
@@ -293,6 +293,12 @@ public function update(User $user, Request $request)
 
         $request->validate(User::rules($user));
         $fields = $request->json()->all();
+        if (($fields['username'] !== $user->getAttribute('username') || isset($fields['password'])) &&
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Api/UserController.php:296]: Please add the corresponding test to validate the following:
- If is an admin user can edit username and password himself
- If is an admin user can edit username and password for other user
- If is a user without permission can not edit
- If is a user with permission can edit himself

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: resources/views/shared/users/sidebar.blade.php [hunk 2]
```diff
@@ -22,35 +22,48 @@
...
-            </small>
-        </div>
-        <div class="form-group">
-            {!! Form::label('password', __('New Password')) !!}
-            <vue-password v-model="formData.password" :disable-toggle=true>
-                <div slot="password-input" slot-scope="props">
-                    {!! Form::password('password', ['id' => 'password', 'rows' => 4, 'class'=> 'form-control', 'v-model'
-                    => 'formData.password', 'autocomplete' => 'new-password', '@input' => 'props.updatePassword($event.target.value)',
-                    'v-bind:class' => '{\'form-control\':true, \'is-invalid\':errors.password}']) !!}
+        <fieldset :disabled="{{ \Auth::user()->hasPermission('edit-user-and-password') || \Auth::user()->is_administrator ? 'false' : 'true' }}">
```
Comment (fagubla) [inline resources/views/shared/users/sidebar.blade.php:25]: Improve this line.

**Review by fagubla — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Api/UserController.php [hunk 3]
```diff
@@ -293,6 +293,11 @@ public function update(User $user, Request $request)
...
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Api/UserController.php:296]: Please update this if to
`if ( (in_array('password' $fields) || in_array('username' $fields)) && `
but I think so the best way is to add a custom middleware
https://laravel.com/docs/10.x/middleware

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Api/UserController.php (hunk 3 above)
Comment (luNunezProcessmaker) [reply ProcessMaker/Http/Controllers/Api/UserController.php:296]: in the first i can change the password, The username will always be in fields, so I check if it is different from the original. 
I am going to study the middleware, in any case I would ask that the estimated time in the ticket be increased

File: resources/views/shared/users/sidebar.blade.php [hunk 4]
```diff
@@ -22,35 +22,48 @@
...
+            <div class="form-group">
+                {!! Form::label('confPassword', __('Confirm Password')) !!}
+                {!! Form::password('confPassword', ['id' => 'confPassword', 'rows' => 4, 'class'=> 'form-control', 'v-model'
+                => 'formData.confPassword', 'autocomplete' => 'new-password', 'v-bind:class' => '{\'form-control\':true, \'is-invalid\':errors.password}']) !!}
+                <div class="invalid-feedback" :style="{display: (errors.password) ? 'block' : 'none' }" role="alert" v-if="errors.password">@{{errors.password[0]}}</div>
+            </div>
+            @cannot('edit-user-and-password')
+                <div class="form-group">
+                    <small class="form-text text-muted">
+                        {{__('To change the current username and password please contact your administrator.')}}
```
Comment (pmPaulis) [inline resources/views/shared/users/sidebar.blade.php:62]: please add this label into translation file

File: resources/views/shared/users/sidebar.blade.php [hunk 5]
```diff
@@ -22,35 +22,48 @@
...
+            </div>
 
-        <div class="form-group">
-            {!! Form::label('confPassword', __('Confirm Password')) !!}
-            {!! Form::password('confPassword', ['id' => 'confPassword', 'rows' => 4, 'class'=> 'form-control', 'v-model'
-            => 'formData.confPassword', 'autocomplete' => 'new-password', 'v-bind:class' => '{\'form-control\':true, \'is-invalid\':errors.password}']) !!}
-            <div class="invalid-feedback" :style="{display: (errors.password) ? 'block' : 'none' }" role="alert" v-if="errors.password">@{{errors.password[0]}}</div>
-        </div>
+            <div class="form-group">
+                {!! Form::label('confPassword', __('Confirm Password')) !!}
```
Comment (pmPaulis) [inline resources/views/shared/users/sidebar.blade.php:54]: please add this label into translation file

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Api/UserController.php (hunk 3 above)
Comment (danloa) [reply ProcessMaker/Http/Controllers/Api/UserController.php:296]: May be you can use Laravel's error-safe Arr::get ( https://laravel.com/docs/8.x/helpers#method-array-get)

File: resources/views/shared/users/sidebar.blade.php (hunk 4 above)
Comment (luNunezProcessmaker) [reply resources/views/shared/users/sidebar.blade.php:62]: this label already exists inside the translation file

File: resources/views/shared/users/sidebar.blade.php (hunk 5 above)
Comment (luNunezProcessmaker) [reply resources/views/shared/users/sidebar.blade.php:54]: this label already exists inside the translation file

**Review by pmPaulis — APPROVED** [review]

**Pull request merged**